File: 23/solution.go
package _23

type ListNode struct {
	Val  int
	Next *ListNode
}

type Heap struct {
	Arr    []*ListNode
	Length int
}

func (h *Heap) push(obj *ListNode) {
	h.Length++
	h.Arr[h.Length] = obj
	h.up(h.Length)
}

func (h *Heap) pop() *ListNode {
	res := h.Arr[1]
	h.Arr[1], h.Arr[h.Length] = h.Arr[h.Length], h.Arr[1]
	h.Length--
	h.down(1)
	return res
}

func (h *Heap) up(idx int) {
	if idx == 1 {
		return
	}
	father_idx := idx / 2
	if h.Arr[idx].Val < h.Arr[father_idx].Val {
		h.Arr[idx], h.Arr[father_idx] = h.Arr[father_idx], h.Arr[idx]
		h.up(father_idx)
	}
}

func (h *Heap) down(idx int) {
	left_idx := idx * 2
	if left_idx <= h.Length && h.Arr[left_idx].Val < h.Arr[idx].Val {
		h.Arr[idx], h.Arr[left_idx] = h.Arr[left_idx], h.Arr[idx]
		h.down(left_idx)
	}
	right_idx := idx*2 + 1
	if right_idx <= h.Length && h.Arr[right_idx].Val < h.Arr[idx].Val {
		h.Arr[idx], h.Arr[right_idx] = h.Arr[right_idx], h.Arr[idx]
		h.down(right_idx)
	}
}

func mergeKLists(lists []*ListNode) *ListNode {
	n := len(lists)
	heap := Heap{make([]*ListNode, n+1), 0}
	for _, item := range lists {
		if item != nil {
			heap.push(item)
		}
	}
	head := &ListNode{-1, nil}
	cur := head
	for heap.Length != 0 {
		tmp := heap.pop()
		cur.Next = tmp
		cur = cur.Next
		if tmp.Next != nil {
			heap.push(tmp.Next)
		}
	}
	return head.Next
}

File: 23/solution.py
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Heapq:
    def __init__(self, n: int):
        self.arr = [None] * (n + 1)
        self.length = 0

    def push(self, obj: ListNode):
        self.length += 1
        self.arr[self.length] = obj
        self.up(self.length)
    
    def pop(self):
        self.arr[1], self.arr[self.length] = self.arr[self.length], self.arr[1]
        res = self.arr[self.length]
        self.length -= 1
        self.down(1)
        return res
    
    def up(self, idx: int):
        if idx == 1:
            return
        father_idx = idx // 2
        if self.arr[idx].val < self.arr[father_idx].val:
            self.arr[idx], self.arr[father_idx] = self.arr[father_idx], self.arr[idx]
            self.up(father_idx)

    def down(self, idx: int):
        left_idx = idx * 2
        if left_idx <= self.length and self.arr[left_idx].val < self.arr[idx].val:
            self.arr[idx], self.arr[left_idx] = self.arr[left_idx], self.arr[idx]
            self.down(left_idx)
        right_idx = idx * 2 + 1
        if right_idx <= self.length and self.arr[right_idx].val < self.arr[idx].val:
            self.arr[idx], self.arr[right_idx] = self.arr[right_idx], self.arr[idx]
            self.down(right_idx)


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = Heapq(len(lists))
        for item in lists:
            if item:
                heap.push(item)
        head = cur = ListNode()
        while heap.length != 0:
            tmp = heap.pop()
            cur.next = tmp
            cur = cur.next
            if tmp.next:
                heap.push(tmp.next)
        return head.next
